python-frontend/src/main/java/org/sonar/python/semantic/v2/types/Assignment.java

```diff
@@ -19,9 +19,86 @@
  */
 package org.sonar.python.semantic.v2.types;
 
+import java.util.ArrayDeque;
+import java.util.Deque;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
 import org.sonar.plugins.python.api.tree.Expression;
 import org.sonar.plugins.python.api.tree.Name;
+import org.sonar.plugins.python.api.tree.Tree;
 import org.sonar.python.semantic.v2.SymbolV2;
+import org.sonar.python.semantic.v2.UsageV2;
+import org.sonar.python.tree.NameImpl;
+import org.sonar.python.types.HasTypeDependencies;
+import org.sonar.python.types.v2.PythonType;
+import org.sonar.python.types.v2.UnionType;
 
-public record Assignment(SymbolV2 lhsSymbol, Name lhsName, Expression rhs) {
+public class Assignment {
+
+  final SymbolV2 lhsSymbol;
+  Name lhsName;
+  Expression rhs;
+  Set<SymbolV2> variableDependencies = new HashSet<>();
+  Set<Assignment> dependents = new HashSet<>();
+  Map<SymbolV2, Set<Assignment>> assignmentsByLhs;
+
+  public Assignment(SymbolV2 lhsSymbol, Name lhsName, Expression rhs, Map<SymbolV2, Set<Assignment>> assignmentsByLhs) {
+    this.lhsSymbol = lhsSymbol;
+    this.lhsName = lhsName;
+    this.rhs = rhs;
+    this.assignmentsByLhs = assignmentsByLhs;
+  }
+
+  void computeDependencies(Expression expression, Set<SymbolV2> trackedVars) {
+    Deque<Expression> workList = new ArrayDeque<>();
+    workList.push(expression);
+    while (!workList.isEmpty()) {
```

Comment:
Add a comment here to rework this part later. Add tests showing type tracking of collection item types

Reply:
I didn't add the comment, as I think we already have some tickets that relate to this anyway. However, I added a test that shows that collection item types actually work (for a trivial case).